designsafe/static/scripts/ng-designsafe/services/file-listing-service.js

```diff
@@ -0,0 +1,632 @@
+import { Subject, ReplaySubject, BehaviorSubject, Observable, from, of, forkJoin, race, throwError } from 'rxjs';
+import { map, tap, switchMap, exhaustMap, take, catchError, share } from 'rxjs/operators';
+import { uuid } from 'uuidv4';
+
+/**
+ * Method to instantiate a subject that subscribes to the observable return
+ * value of the last callback projected to it and cancels the result of any
+ * previous calls. The callback must be a function that returns an observable.
+ */
+const takeLatestSubscriber = () => {
+    let subject = new Subject();
+    subject = subject.pipe(
+        switchMap((callback) => callback()),
+        share() // Allow subscription to in-flight http requests.
+    );
+    subject.subscribe();
+    return subject;
+};
+
+/**
+ * Method to instantiate a subject that subscribes to the observable return
+ * value of the first callback passed to it and ignores anything new projected
+ * to it until that observable resolves. The callback must be a function that
+ * returns an observable.
+ */
+const takeLeadingSubscriber = () => {
+    const subject = new Subject();
+    subject
+        .pipe(
+            exhaustMap((callback) => callback()),
+            share() // Allow subscription to in-flight http requests.
+        )
+        .subscribe();
+    return subject;
+};
```

Comment:
These are repeated in `file-operation-service.js` - can we consolidate? Might they be used by future services in the future?

Reply:
Good call, moving them into an `_rxjs-utils.js` in the top-level service module.